### PR TITLE
feat: add plugin audit

### PR DESCRIPTION
**Pull request by Dav-14**

### CODE REVIEW

File: components/gateway/Caddyfile [hunk 1]
```diff
@@ -101,13 +105,33 @@
 	}
 }
 
+(audit) {
```
Comment (paul-nicolas) [inline components/gateway/Caddyfile:108]: Should we add it also for /versions and all other api calls returning 502 ?

File: components/gateway/pkg/plugins/audit.go [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+	}
+
+	return conn, nil
+}
+
+func newNatsPublisherWithConn(conn *nats.Conn, logger watermill.LoggerAdapter, config wNats.PublisherConfig) (*wNats.Publisher, error) {
+	return wNats.NewPublisherWithNatsConn(conn, config.GetPublisherPublishConfig(), logger)
+}
+
+func (a *Audit) provisionNatsPublisher() error {
```
Comment (paul-nicolas) [inline components/gateway/pkg/plugins/audit.go:202]: Should those functions about instantiating NATS be in go-libs ? (Just asking for opinions here, I'm fine with either letting this code here or putting it in golibs)

File: components/gateway/pkg/plugins/audit.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+	buf.Reset()
+	defer a.bufPool.Put(buf)
+
+	rww := NewResponseWriterWrapper(w, buf)
+	if err := next.ServeHTTP(rww, r); err != nil {
+		return err
+	}
+
+	response := messages.NewHttpResponse(
+		*rww.statusCode,
```
Comment (paul-nicolas) [inline components/gateway/pkg/plugins/audit.go:330]: I have a doubt here, should we check that status code is never nil ?

Comment (paul-nicolas) [reply components/gateway/pkg/plugins/audit.go:330]: Nevermind, there is a default inside the wrapper

**Review by paul-nicolas — APPROVED** [review]

File: components/gateway/Caddyfile (hunk 1 above)
Comment (gfyrag) [reply components/gateway/Caddyfile:108]: I think we should log every requests.
@flemzord ?

File: components/gateway/pkg/plugins/audit.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+
+	return res, nil
+}
+
+// Implements the caddy.Provisioner interface.
+func (a *Audit) Provision(ctx caddy.Context) error {
+	// We have to instantiate a different logger here instead of using
+	// caddy's logger because we need a different interface for the watermill
+	// logger.
+	a.logger = logging.NewLogrus(logrus.New())
```
Comment (gfyrag) [inline components/gateway/pkg/plugins/audit.go:169]: Can you implement the watermil logger interface using the caddy logger as underlying logger?

File: components/gateway/pkg/plugins/audit.go (hunk 2 above)
Comment (gfyrag) [reply components/gateway/pkg/plugins/audit.go:202]: Yes, it smells like copy/paste no?

File: components/gateway/pkg/plugins/audit.go [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+		Body:   "",
+	}
+
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		if !errors.Is(err, io.EOF) {
+			return err
+		}
+	}
+
```
Comment (gfyrag) [inline components/gateway/pkg/plugins/audit.go:313]: You need to close the original body before replacing it.

File: components/gateway/Caddyfile (hunk 1 above)
Comment (flemzord) [reply components/gateway/Caddyfile:108]: All calls must be logged

**Review by paul-nicolas — APPROVED** [review]

**Review by gfyrag — APPROVED** [review]

**Pull request merged**